src/utils.c

```diff
@@ -110,30 +117,231 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+      }
+    } else {
+      SEXP notricols = ricols;
+      SEXP ricols = PROTECT(allocVector(INTSXP, nx-nc)); protecti++;
+      int *icols = INTEGER(ricols);
+      error("todo colnamesInt inverse isInteger|isReal");
+      //for (int i=0; i<nc; i++) {
+      //  if ((icols[i]>nx) || (icols[i]<1))
+      //    error(_("argument specifying columns specify non existing column(s): cols[%d]=%d"), i+1, icols[i]); // handles NAs also
+      //}
     }
   } else if (isString(cols)) {
-    SEXP xnames = PROTECT(getAttrib(x, R_NamesSymbol)); protecti++;
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
     if (isNull(xnames))
       error(_("'x' argument data.table has no names"));
-    ricols = PROTECT(chmatch(cols, xnames, 0)); protecti++;
-    int *icols = INTEGER(ricols);
+    if (!binverse) {
+      ricols = PROTECT(chmatch(cols, xnames, 0)); protecti++;
+      int *icols = INTEGER(ricols);
+      for (int i=0; i<nc; i++) {
+        if (icols[i]==0)
+          error(_("argument specifying columns specify non existing column(s): cols[%d]='%s'"), i+1, CHAR(STRING_ELT(cols, i))); // handles NAs also
+      }
+    } else {
+      error("todo colnamesInt inverse isString");
+      /*SEXP xmatch = PROTECT(chmatch(xnames, cols, 0)); protecti++; // we actually need a no-match
+      int *xmatchp = INTEGER(xmatch);
+      int len = 0;
+      for (int i=0; i<length(xmatch); i++) {
+        if (xmatchp[i]==0)
+          len++; // TODO as tests for NAs!
+      }
+      ricols = PROTECT(allocVector(INTSXP, len)); protecti++;*/
+    }
+  } else if (isLogical(cols)) {
+    int *lcols = LOGICAL(cols);
+    ricols = PROTECT(allocVector(INTSXP, nc)); protecti++;
+    int *icols = LOGICAL(ricols);
+    int ntrue = 0;
     for (int i=0; i<nc; i++) {
-      if (icols[i]==0)
-        error(_("argument specifying columns specify non existing column(s): cols[%d]='%s'"), i+1, CHAR(STRING_ELT(cols, i))); // handles NAs also
+      if (lcols[i]==NA_LOGICAL)
+        error(_("argument specifying columns is logical and has NAs"));
+      else if ((!binverse && lcols[i]) || (binverse && !lcols[i]))
+        icols[ntrue++] = i+1;
     }
+    SETLENGTH(ricols, ntrue);
   } else {
-    error(_("argument specifying columns must be character or numeric"));
+    error(_("argument specifying columns must be character, numeric or logical"));
   }
   if (LOGICAL(check_dups)[0] && any_duplicated(ricols, FALSE))
     error(_("argument specifying columns specify duplicated column(s)"));
   UNPROTECT(protecti);
   return ricols;
 }
 
+/* replace_dot_alias
+ * we don't just simply alias .=list because i) list is a primitive (faster to iterate) and ii) we test for use
+ * of "list" in several places so it saves having to remember to write "." || "list" in those places
+ * replace_dot_aliasR wrapper should be used to prevent update of input
+ */
+SEXP replace_dot_alias(SEXP x) {
+  if (isLanguage(x) && !isFunction(CAR(x))) {
+    if (CAR(x) == sym_bquote) // handling `.` inside bquote, #1912
```

Comment:
nice (though a bit surprising) that `==` works here!

Reply:
Maybe because it is not STRSXP but CHARSXP